### ISSUE TITLE
JSON body with very large numbers causes problems

### ISSUE THREAD

**Issue by dstockto** [zfcampus/zf-content-validation#57]
If an input body into an apigility API contains a field with a very large number, json_encode will emit a notice:

    Notice: json_decode(): integer overflow detected in /vagrant_web/vendor/zfcampus/zf-content-negotiation/src/ContentTypeListener.php on line 172

It doesn't matter if the field is defined in the API or not, it will happen on any POST, PUT or PATCH. 

    { "foobar": 128934701823471023487128390871 }

Since the decoding happens before any validation occurs, this is understandable. Also, unfortunately, it appears that tacking on the silence operator (@) to hide the notice actually changes the behavior. As the code sits now, the validators and the resource code will receive 9223372036854775807 for the value of those large integer containing fields.

If I send in the large number for a field with a StringLength validator, after the Notice and stack trace, I get a json body with the validator indicating that the field expected a string but didn't get that.

If I change line 172 to be 

    $data = @json_decode($json, true);

Then I don't see the notice and I'm getting the same (expected) validation errors. But the input number is smashed down to PHP_INT_MAX. There's an option on json_decode called JSON_BIGINT_AS_STRING which seems it should be able to use the large number as a string (at least that's how I read it) but it doesn't seem to have any effect whatsoever.

I'm not really happy with the @ solution since @ is evil and all that, but since JSON_BIGINT_AS_STRING doesn't appear to work and there doesn't appear to be any way at all to get the actual input in any format, I'm not sure if anything can be done.

TL;DR - If anything can be done about this that would ideally not spew a notice but could be dealt with in a way that would give back some sort of nice error (that would be JSON at the end) that would be ideal. I don't know if it's possible though.

**Comment by Wilt** [comment]
I am not sure if this is an Apigility issue.

PHP has maximum of integer sized that it can handle, check also [here](http://stackoverflow.com/questions/670662/whats-the-maximum-size-for-an-int-in-php) or [here](http://php.net/manual/en/language.types.integer.php) for more on this topic.

32-bit builds of PHP:
min: -2,147,483,648  
max: 2,147,483,647

64-bit builds of PHP:
min: -9,223,372,036,854,775,808 
max: 9,223,372,036,854,775,807

Your 128934701823471023487128390871 exceeds this more then a little. Try posting it as a string and handle it internally in PHP with special functionality like [proposed here](http://stackoverflow.com/a/211548/1697459).

**Comment by dstockto** [comment]
@Wilt I tried to patch zf-content-validation with that flag (for handling large numbers as strings) back before I opened this ticket. It didn't appear to make any difference at all.

I agree that the input is way bigger than a normal int that PHP can handle, however we should be able to ensure that any input into our APIs won't crash them.

**Comment by weierophinney** [comment]
I think this should likely be against zf-content-negotiation, as that's where the de-serialization happens. Could you re-open this over there, please, @dstockto ?

**Comment by dstockto** [comment]
I've moved it over and referenced this from there. Closing out this one.
